db/migrations/000001_db_setup.up.sql

```diff
@@ -0,0 +1,178 @@
+CREATE TYPE
+    "order_status" AS ENUM (
+        'pending',
+        'paid',
+        'shipped',
+        'delivered',
+        'cancelled'
+    );
+
+CREATE TYPE
+    "coupon_type" AS ENUM (
+        'percentage',
+        'fixed',
+        'shipping'
+    );
+
+CREATE TYPE "role_type" AS ENUM ( 'admin', 'customer' );
+
+CREATE TABLE
+    "cart" (
+        "id" SERIAL PRIMARY KEY,
+        "user_id" INT NOT NULL,
+        "shop_id" INT NOT NULL
+    );
+
+CREATE TABLE
+    "cart_product" (
+        "cart_id" INT NOT NULL,
+        "product_id" INT NOT NULL,
+        "quantity" INT NOT NULL
+    );
+
+CREATE TABLE
+    "order_history" (
+        "id" SERIAL PRIMARY KEY,
+        "user_id" INT NOT NULL,
+        "shop_id" INT NOT NULL,
+        "shipment" INT NOT NULL,
+        "total_price" INT NOT NULL,
+        "status" order_status NOT NULL,
+        "created_at" TIMESTAMPTZ NOT NULL DEFAULT NOW()
+    );
+
+CREATE TABLE
+    "order_detail" (
+        "order_id" INT NOT NULL,
+        "product_id" INT NOT NULL,
+        "product_version" INT NOT NULL,
+        "quantity" INT NOT NULL
+    );
+
+CREATE TABLE
+    "cart_coupon" (
+        "cart_id" INT NOT NULL,
+        "coupon_id" INT NOT NULL
+    );
+
+CREATE TABLE
+    "product" (
+        "id" SERIAL PRIMARY KEY,
+        "version" INT NOT NULL DEFAULT 1,
+        "shop_id" INT NOT NULL,
+        "name" VARCHAR(255) NOT NULL,
+        "description" TEXT NOT NULL,
+        "price" DECIMAL(10, 2) NOT NULL,
+        "image_id" UUID NOT NULL,
+        "due_date" TIMESTAMPTZ NOT NULL,
+        "stock" INT NOT NULL,
+        "sales" INT NOT NULL,
+        "enabled" BOOLEAN NOT NULL DEFAULT TRUE,
+        UNIQUE ("id", "version")
+    );
```

Comment:
After some experiment, we can't have duplicate primary key in a table 🦦
Here are two solution:
keep this table and add another table to store all history product (?include newest)
or
remove primary key constraint

Reply:
is [composite primary key](https://www.commandprompt.com/education/composite-primary-keys-in-postgresql/) a good solution?

Reply:
also if we are going to use this solution, `id` probably shouldn't be `SERIAL` as well

Reply:
But using a composite primary key means that other tables must use the same combination of columns as a composite foreign key when they need to reference it.
And most of time we only need product id.

Reply:
I prefer first one.

Reply:
i have no idea what i'm doing btw

Reply:
How about using int32, where the first 24 bits represent the product ID, and the remaining 8 bits represent the version value, combining them as a 32-bit primary key?

Reply:
> what if we use composite primary keys and break up 💔 foreign key relationships? because wouldn't it be weird if your foreign key didn't change but the underlying row has changed?

That's fine?

Reply:
```suggestion
CREATE TABLE
    "product" (
        "id" SERIAL PRIMARY KEY,
        "version" INT NOT NULL DEFAULT 1,
        "shop_id" INT NOT NULL,
        "name" VARCHAR(255) NOT NULL,
        "description" TEXT NOT NULL,
        "price" DECIMAL(10, 2) NOT NULL,
        "image_id" UUID NOT NULL,
        "due_date" TIMESTAMPTZ NOT NULL,
        "stock" INT NOT NULL,
        "sales" INT NOT NULL,
        "enabled" BOOLEAN NOT NULL DEFAULT TRUE,
        UNIQUE ("id", "version")
    );
CREATE TABLE
    "product_version_history"(
        "id" INT NOT NULL,
        "version" INT NOT NULL DEFAULT 1,
        "shop_id" INT NOT NULL,
        "name" VARCHAR(255) NOT NULL,
        "description" TEXT NOT NULL,
        "price" DECIMAL(10, 2) NOT NULL,
        "image_id" UUID NOT NULL,
        "due_date" TIMESTAMPTZ NOT NULL,
        "stock" INT NOT NULL,
        "sales" INT NOT NULL,
        "enabled" BOOLEAN NOT NULL DEFAULT TRUE,
        UNIQUE ("id", "version")
    )
```

Reply:
In [🐹](https://github.com/jykuo-love-shiritori/twp/pull/24#discussion_r1381634644) case, we can do below operation to "product"  table only
1. modify existed product
2. insert a new product
3. delete a disabled product

and after 1, 2 operation, add a current version copy into "product_version_history"

Reply:
> How about using int32, where the first 24 bits represent the product ID, and the remaining 8 bits represent the version value, combining them as a 32-bit primary key?

stop it, get some help